elastic/beats
# [filebeat][httpjson] Initial implementation of httpjson v2

## What does this PR do?

Creates an initial implementation of the new Httpjson input.

## Why is it important?

HTTPJSON is too specific to our particular modules use cases and is hard to extend or reuse out of them. For this we wanted to do an effort to generalize the input so it can be used in a more flexible way.

## Checklist


## Author's Checklist

- [x] The possible values that a transform modifies are defined by where it is configured
- [x] Load cursor from registry
- [x] Store cursor in registry
- [x] Custom cursor values
- [x] Supports oauth
- [x] Supports basic auth
- [x] Supports automatic retries
- [x] Rate limit with custom values
- [x] Values can have a default in case of empty or failure
- [x] Value templates use predefined helper functions
  - [x] now
  - [x] parseDuration
  - [x] parseDate
  - [x] formatDate
  - [x] parseTimestamp
  - [x] parseTimestampMilli
  - [x] parseTimestamlNano
  - [x] getRFC5988Link
- [x] Request transforms
  - [x] Set
  - [x] Append
- [x] Response transforms
  - [x] Set
  - [x] Append
  - [x] Split
    - [x] Arrays
    - [x] Maps
- [x] Response pagination
  - [x] Set
  - [x] Append
- [x] Redirects can forward headers
- [x] Documentation for the new input
- [x] Update old documentation to mention deprecation and migration steps
- [x] Add missing tests

 
## How to test this PR locally

- Set a mock endpoint to respond to http requests
- Set up a config that uses the new input

```yml
- type: httpjson
  is_v2: true
  interval: 10s
  request.method: post
  request.url: https://example.com
  request.transforms:
    - set:
        target: body.foo
        value: bazz
  response.transforms:
    - delete:
        target: body.this_is_private
  response.pagination:
    - set:
        target: url.value
        value: "https://example.com/page"
    - set:
        target: url.params.p
        value: "{{.last_response.body.page}}"

output.console:
  pretty: true

```

Then in `x-pack/filebeat` run `mage build && ./filebeat -e -v` and check the output.

## Related issues

- Closes https://github.com/elastic/beats/issues/21465
- Closes https://github.com/elastic/beats/issues/20559
- Closes https://github.com/elastic/beats/issues/19486
- Relates https://github.com/elastic/beats/issues/20379
